src/System/Console/GetOpt/Generics.hs

```diff
@@ -111,51 +116,83 @@ processFields :: forall a xs .
   (Generic a, Code a ~ '[xs], SingI xs, All Option xs) =>
   String -> Modifiers -> [String] -> NP FieldInfo xs -> Result a
 processFields header modifiers args fields =
-    helpWrapper header modifiers args fields *>
+    mkInitialFieldStates modifiers fields >>= \ initialFieldStates ->
+
+    showHelp *>
+
     let (options, arguments, parseErrors) =
           getOpt Permute (mkOptDescrs modifiers fields) args
     in
 
-    -- report parse errors
-    (case parseErrors of
-      [] -> pure ()
-      errs -> Errors errs) *>
+    reportParseErrors parseErrors *>
 
-    -- report unknown arguments
-    (case arguments of
+    reportInvalidPositionalArguments arguments *>
+
+    produceResult initialFieldStates options arguments
+  where
+    showHelp :: Result ()
+    showHelp = helpWrapper header modifiers args fields
+
+    reportParseErrors :: [String] -> Result ()
+    reportParseErrors parseErrors = case parseErrors of
       [] -> pure ()
-      _ -> Errors (map ("unknown argument: " ++) arguments)) *>
+      errs -> Errors errs
+
+    reportInvalidPositionalArguments :: [String] -> Result ()
+    reportInvalidPositionalArguments arguments =
+      when (not $ hasPositionalArgumentsField modifiers) $
+        case arguments of
+          [] -> pure ()
+          _ -> Errors (map ("unknown argument: " ++) arguments)
+
+    produceResult :: NP FieldState xs -> [NS FieldState xs] -> [String] -> Result a
+    produceResult initialFieldStates options arguments =
+      (to . SOP . Z) <$>
+        collectResult arguments (project options initialFieldStates)
 
-    ((to . SOP . Z) <$>
-      collectErrors (project options (mkEmptyArguments fields)))
 
 mkOptDescrs :: forall xs . All Option xs =>
   Modifiers -> NP FieldInfo xs -> [OptDescr (NS FieldState xs)]
 mkOptDescrs modifiers fields =
-  map toOptDescr $ sumList $ npMap (mkOptDescr modifiers) fields
+  mapMaybe toOptDescr $ sumList $ npMap (mkOptDescr modifiers) fields
 
-newtype OptDescrE a = OptDescrE (OptDescr (FieldState a))
+newtype OptDescrE a = OptDescrE (Maybe (OptDescr (FieldState a)))
 
 mkOptDescr :: forall a . Option a => Modifiers -> FieldInfo a -> OptDescrE a
 mkOptDescr modifiers (FieldInfo name) = OptDescrE $
-  Option
-    (mkShortOptions modifiers name)
-    [mkLongOption modifiers name]
-    _toOption
-    ""
-
-toOptDescr :: NS OptDescrE xs -> OptDescr (NS FieldState xs)
-toOptDescr (Z (OptDescrE a)) = fmap Z a
-toOptDescr (S a) = fmap S (toOptDescr a)
-
-mkEmptyArguments :: forall xs . (SingI xs, All Option xs) =>
-  NP FieldInfo xs -> NP FieldState xs
-mkEmptyArguments fields = case (sing :: Sing xs, fields) of
-  (SNil, Nil) -> Nil
+  if isPositionalArgumentsField modifiers name
```

Comment:
It would be `guard $ not $ isPositionalArgumentsField ...` which I find less clear.